pkg/devicerepository/store/bleve/store.go

```diff
@@ -0,0 +1,192 @@
+// Copyright © 2020 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bleve
+
+import (
+	"github.com/blevesearch/bleve"
+	"github.com/blevesearch/bleve/search/query"
+	"go.thethings.network/lorawan-stack/v3/pkg/devicerepository/store"
+	"go.thethings.network/lorawan-stack/v3/pkg/errors"
+	"go.thethings.network/lorawan-stack/v3/pkg/jsonpb"
+	"go.thethings.network/lorawan-stack/v3/pkg/ttnpb"
+)
+
+var (
+	errCorruptedIndex = errors.DefineCorruption("corrupted_index", "corrupted index file")
+)
+
+// GetBrands lists available end device vendors.
+func (s *bleveStore) GetBrands(req store.GetBrandsRequest) (*store.GetBrandsResponse, error) {
+	queries := []query.Query{
+		bleve.NewMatchAllQuery(),
+	}
+	if q := req.Search; q != "" {
+		queries = append(queries, bleve.NewQueryStringQuery(q))
+	}
+	if q := req.BrandID; q != "" {
+		query := bleve.NewMatchQuery(q)
+		query.SetField("BrandID")
+		queries = append(queries, query)
+	}
+
+	searchRequest := bleve.NewSearchRequest(bleve.NewConjunctionQuery(queries...))
+	searchRequest.Size = int(req.Limit)
+	if req.Page == 0 {
+		req.Page = 1
+	}
+	searchRequest.From = int((req.Page - 1) * req.Limit)
+
+	searchRequest.Fields = []string{"BrandPB"}
```

Comment:
So this is a full-text search on JSON, or is Bleve smart enough to look at field values?

Reply:
This is only the list of fields that bleve will return for documents that match the query, not the list of fields on which we search. Bleve will search all fields.